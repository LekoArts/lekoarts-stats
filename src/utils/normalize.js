import _ from 'lodash'

export function flatten(data) {
  return data.flatMap((g) => {
    return g.repos.map((r) => {
      return {
        name: r.name,
        datetime: g.datetime,
        stars: r.stars,
        forks: r.forks,
        url: r.url,
      }
    })
  })
}

/**
 * _.mapValues: https://lodash.com/docs/4.17.15#mapValues
 * _.groupBy: https://lodash.com/docs/4.17.15#groupBy
 * _.omit: https://lodash.com/docs/4.17.15#omit
 */

export function constructShape(flat, name, omit) {
  return _.mapValues(_.groupBy(flat, name), (list) => list.map((entry) => _.omit(entry, omit)))
}

export function normalizeGithub(data, name, omit) {
  const flat = flatten(data)
  return constructShape(flat, name, omit)
}

export function nivoGithubFormatter(data, name) {
  const keys = Object.keys(data)
  const nivoData = []

  for (const key of keys) {
    const entry = data[key]

    const values = entry.map((e) => ({
      x: e.datetime,
      y: e[name],
    }))

    const obj = {
      id: key,
      data: values,
    }

    nivoData.push(obj)
  }

  return nivoData.reverse()
}

export function nivoTwitterFormatter(data, name) {
  const values = data.map((e) => ({
    x: e.datetime,
    y: e[name],
  }))

  return [
    {
      id: name,
      data: values,
    },
  ]
}
