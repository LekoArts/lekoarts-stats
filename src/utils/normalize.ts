import { groupBy, mapValues, omit } from 'lodash-es'

/**
 * _.mapValues: https://lodash.com/docs/4.17.15#mapValues
 * _.groupBy: https://lodash.com/docs/4.17.15#groupBy
 * _.omit: https://lodash.com/docs/4.17.15#omit
 */

/**
 * Formats 2020-07-18 14:02:07.078+00 to 2020-07-18
 */
export function formatCreatedAt(date: string) {
  return date.split(' ')[0]
}

interface IGitHubEntry {
  id: number
  name: string
  createdAt: string
  forks: number
  stars: number
}

export function normalizeGithub(data: IGitHubEntry[], name: keyof IGitHubEntry) {
  return mapValues(groupBy(data, name), list => list.map(entry => omit(entry, name))) as GitHubFormatterDataInput
}

interface GitHubFormatterDataInput {
  [key: string]: Omit<IGitHubEntry, 'name'>[]
}

export function nivoGithubFormatter(data: GitHubFormatterDataInput, name: keyof Omit<IGitHubEntry, 'name'>) {
  const keys = Object.keys(data)
  const nivoData = []

  for (const key of keys) {
    const entry = data[key]

    const values = entry.map(e => ({
      x: formatCreatedAt(e.createdAt),
      y: e[name],
    }))

    const obj = {
      id: key,
      data: values,
    }

    nivoData.push(obj)
  }

  return nivoData.reverse()
}

interface ITwitterEntry {
  id: number
  createdAt: string
  tweets: number
  followers: number
}

export function nivoTwitterFormatter(data: ITwitterEntry[]) {
  const followersValues = data.map(e => ({
    x: formatCreatedAt(e.createdAt),
    y: e.followers,
  }))
  const tweetsValues = data.map(e => ({
    x: formatCreatedAt(e.createdAt),
    y: e.tweets,
  }))

  return [
    {
      id: 'followers',
      data: followersValues,
    },
    {
      id: 'tweets',
      data: tweetsValues,
    },
  ]
}

interface IMastodonEntry {
  id: number
  createdAt: string
  followersCount: number
  tootsCount: number
}

export function nivoMastodonFormatter(data: IMastodonEntry[]) {
  const followersValues = data.map(e => ({
    x: formatCreatedAt(e.createdAt),
    y: e.followersCount,
  }))
  const tweetsValues = data.map(e => ({
    x: formatCreatedAt(e.createdAt),
    y: e.tootsCount,
  }))

  return [
    {
      id: 'followers',
      data: followersValues,
    },
    {
      id: 'toots',
      data: tweetsValues,
    },
  ]
}

interface ITraktEntry {
  id: number
  createdAt: string
  moviesWatched: number
  showsWatched: number
  episodesWatched: number
  ratings: number
}

export function nivoTraktFormatter(data: ITraktEntry[], name: keyof ITraktEntry) {
  const values = data.map(e => ({
    x: formatCreatedAt(e.createdAt),
    y: e[name],
  }))

  return [
    {
      id: name,
      data: values,
    },
  ]
}
