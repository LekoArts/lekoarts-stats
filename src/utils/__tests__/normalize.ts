import { nivoGithubFormatter, nivoTwitterFormatter, normalizeGithub } from '../normalize'
import { data as githubData } from './__fixtures__/github'
import { data as twitterData } from './__fixtures__/twitter'

describe('normalizeGithub', () => {
  it('should flatten and group data by name', () => {
    const normalized = normalizeGithub(githubData, 'name')
    const namedEntry = normalized['gatsby-starter-minimal-blog']

    expect(normalized).toMatchSnapshot()
    expect(namedEntry).toStrictEqual([
      {
        createdAt: '2020-07-17',
        forks: 154,
        stars: 517,
        id: '2',
      },
      {
        createdAt: '2020-07-18',
        forks: 153,
        stars: 517,
        id: '5',
      },
    ])
  })
})

describe('nivoGithubFormatter', () => {
  it('should output correct shape via input', () => {
    const normalized = normalizeGithub(githubData, 'name')
    const nivo = nivoGithubFormatter(normalized, 'stars')

    const objectShape = [
      {
        data: [
          {
            x: '2020-07-17',
            y: 517,
          },
          {
            x: '2020-07-18',
            y: 517,
          },
        ],
        id: 'gatsby-starter-minimal-blog',
      },
    ]

    expect(nivo).toMatchSnapshot()
    expect(nivo).toEqual(expect.arrayContaining(objectShape))
  })
})

describe('nivoTwitterFormatter', () => {
  it('should output correct shape via input', () => {
    const nivo = nivoTwitterFormatter(twitterData, 'followers')

    const objectShape = [
      {
        data: [
          {
            x: '2020-07-16',
            y: 1126,
          },
          {
            x: '2020-07-17',
            y: 1137,
          },
          {
            x: '2020-07-18',
            y: 1150,
          },
        ],
        id: 'followers',
      },
    ]

    expect(nivo).toMatchSnapshot()
    expect(nivo).toEqual(expect.arrayContaining(objectShape))
  })
})
