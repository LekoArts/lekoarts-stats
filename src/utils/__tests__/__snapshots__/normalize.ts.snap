// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`nivoBlueskyFormatter > should output correct shape via input 1`] = `
[
  {
    "data": [
      {
        "x": "2020-07-16",
        "y": 1126,
      },
      {
        "x": "2020-07-17",
        "y": 1137,
      },
      {
        "x": "2020-07-18",
        "y": 1150,
      },
    ],
    "id": "followers",
  },
  {
    "data": [
      {
        "x": "2020-07-16",
        "y": 880,
      },
      {
        "x": "2020-07-17",
        "y": 882,
      },
      {
        "x": "2020-07-18",
        "y": 890,
      },
    ],
    "id": "posts",
  },
]
`;

exports[`nivoGithubFormatter > should output correct shape via input 1`] = `
[
  {
    "data": [
      {
        "x": "2020-07-17",
        "y": 113,
      },
      {
        "x": "2020-07-18",
        "y": 113,
      },
    ],
    "id": "gatsby-starter-portfolio-bella",
  },
  {
    "data": [
      {
        "x": "2020-07-17",
        "y": 517,
      },
      {
        "x": "2020-07-18",
        "y": 517,
      },
    ],
    "id": "gatsby-starter-minimal-blog",
  },
  {
    "data": [
      {
        "x": "2020-07-17",
        "y": 745,
      },
      {
        "x": "2020-07-18",
        "y": 749,
      },
    ],
    "id": "gatsby-themes",
  },
]
`;

exports[`nivoMastodonFormatter > should output correct shape via input 1`] = `
[
  {
    "data": [
      {
        "x": "2020-07-16",
        "y": 1126,
      },
      {
        "x": "2020-07-17",
        "y": 1137,
      },
      {
        "x": "2020-07-18",
        "y": 1150,
      },
    ],
    "id": "followers",
  },
  {
    "data": [
      {
        "x": "2020-07-16",
        "y": 880,
      },
      {
        "x": "2020-07-17",
        "y": 882,
      },
      {
        "x": "2020-07-18",
        "y": 890,
      },
    ],
    "id": "toots",
  },
]
`;

exports[`nivoTraktFormatter > should output correct shape via input 1`] = `
[
  {
    "data": [
      {
        "x": "2020-07-16",
        "y": 380,
      },
      {
        "x": "2020-07-17",
        "y": 375,
      },
    ],
    "id": "moviesWatched",
  },
]
`;

exports[`nivoTwitterFormatter > should output correct shape via input 1`] = `
[
  {
    "data": [
      {
        "x": "2020-07-16",
        "y": 1126,
      },
      {
        "x": "2020-07-17",
        "y": 1137,
      },
      {
        "x": "2020-07-18",
        "y": 1150,
      },
    ],
    "id": "followers",
  },
  {
    "data": [
      {
        "x": "2020-07-16",
        "y": 880,
      },
      {
        "x": "2020-07-17",
        "y": 882,
      },
      {
        "x": "2020-07-18",
        "y": 890,
      },
    ],
    "id": "tweets",
  },
]
`;

exports[`normalizeGithub > should flatten and group data by name 1`] = `
{
  "gatsby-starter-minimal-blog": [
    {
      "createdAt": "2020-07-17 14:02:07.078+00",
      "forks": 154,
      "id": 2,
      "stars": 517,
    },
    {
      "createdAt": "2020-07-18 14:02:07.078+00",
      "forks": 153,
      "id": 5,
      "stars": 517,
    },
  ],
  "gatsby-starter-portfolio-bella": [
    {
      "createdAt": "2020-07-17 14:02:07.078+00",
      "forks": 19,
      "id": 3,
      "stars": 113,
    },
    {
      "createdAt": "2020-07-18 14:02:07.078+00",
      "forks": 18,
      "id": 6,
      "stars": 113,
    },
  ],
  "gatsby-themes": [
    {
      "createdAt": "2020-07-17 14:02:07.078+00",
      "forks": 243,
      "id": 1,
      "stars": 745,
    },
    {
      "createdAt": "2020-07-18 14:02:07.078+00",
      "forks": 245,
      "id": 4,
      "stars": 749,
    },
  ],
}
`;
